### PR TITLE
Add advanced forecasting notebook and climpred multi-year statistics verification

### LINKED ISSUE #240: Advanced forecasting notebook
## Description

Notebook runs RAVEN through WPS in probabilistic and deterministic forecast mode and displays results in graphics, then compares results with a climatology-based ensemble forecast.

### LINKED ISSUE #255: WPS wrapping of algorithms to evaluate long-term forecasting ability (multi-year statistics)
## Description

Describe what you were trying to get done or your feature request.

## Environment

* Raven version used, if any:
* Python version, if any:
* Operating System:

## Steps to Reproduce

```
Paste the command(s) you ran and the output.
```

## Additional Information

Links to other issues or sources.

### PR DESCRIPTION
**Pull request by richardarsenault**
## Overview

This PR adds a notebook to compare climatology ESP and operational forecasting methods on the same day (closes #240).

At the same time, tools for multi-year and multi-initial-date forecast verification are added (uses the climpred functionalities) in the ravenpy repo (PR 78). Some paths might need to be changed, and the test data in the raven-testdata repo, branch "add_LSJ_ERA5" should be merged and used directly. 

This PR also fixes #255

### CODE REVIEW

**Comment by cjauvin** [comment]
Richard could you please try to apply the `pre-commit` tool that I made you install please? 
```pre-commit run --all-files ```
to run it manually (normally you don't have to do that, as the git hook does it automatically).

**Comment by Zeitsperre** [comment]
@richardarsenault @cjauvin There isn't a precommit setup here. I can add one But I wouldn't suggest using it on this PR as the changes will be enormous. Opening a PR for this right now.

**Comment by cjauvin** [comment]
ah yes sorry my bad guys..

**Comment by richardarsenault** [comment]
Ok no problem! @cjauvin is it ok for the current PR to remain as-is?

**Comment by cjauvin** [comment]
@richardarsenault yes it's ok, we'll do it after it's merged.

**Comment by Zeitsperre** [comment]
Also, don't fret over the failing docs build. This is due to an API change now addressed in #335.

File: raven/processes/wps_hindcast.py
Comment (huard) [inline raven/processes/wps_hindcast.py:69]: @richardarsenault I've commented this bit and the test still works. I thought you had issues with this ?

Comment (richardarsenault) [reply raven/processes/wps_hindcast.py:69]: Yes, the test still passes, but the rvc is not being used. The results are different. If we comment it out, the model starts with an empty rvc file.

Comment (huard) [reply raven/processes/wps_hindcast.py:69]: Ok, so I'll add an assert to test the result and then try to trace back why the model is not initialized where it should.

Comment (richardarsenault) [reply raven/processes/wps_hindcast.py:69]: OK good. Some info/notes that I took:
Problem is in RavenProcess.initialize.
Base.py
rvc = self.get_config(request, ids=("rvc",))["rvc"]
rvc n'est pas dans la request ?! mais est dans self...

Comment (huard) [reply raven/processes/wps_hindcast.py:69]: Maintenant réglé. là j'ai des problèmes avec climatology_esp. Ça marchait de ton bord ? Avec quelle version de ravenpy?

**Comment by huard** [comment]
@richardarsenault Please review my latest changes, take another look at the notebook to clarify the first paragraph and this is ready to merge.

**Review by huard — APPROVED** [review]

**Comment by richardarsenault** [comment]
Oui ça marchait avec 0.2.3, je vais refaire des tests avec la nouvelle version si tu veux

**Comment by richardarsenault** [comment]
Ok I will review it now, forget my last message

**Comment by richardarsenault** [comment]
OK so I am not sure exactly why, but all my tests were failing after pulling the branch. Probably due to the other numerous updates? After git pull-ing the ravenpy, raven-testdata and raven-wps masters and branches etc, Tests are still failing and notebook not running. So I am not sure what I have done, but I recomment we merge this if it works on your end. I'll then update everything from the master and see if I can make things work. 

Has the raven binary changed at all? Any other special steps I should take since we packaged ravenpy and ravenC in conda, and merged #317?

**Comment by Zeitsperre** [comment]
@richardarsenault 

The method for installing the newest raven with `ravenc`, `ostrich`, and `RavenPy` is different now, yes. The method that uses pip install in the `makefile` needs to be updated since we no longer need to the `--install-options` if we leverage `conda`. My suggestion would be to create a clean env using the `environment.yml` and try running your tests again.

For the notebook, if you are receiving any process outputs as GeoJSON or GeoTIFF objects, the new `Birdy` has changed how these are read into the notebooks. GeoJSONs are now read like dictionaries (loaded with `geojson.loads`), and `GeoTIFF` are accessed as `xarray` objects (where the variable is `band`; therefore `ds.where(band=1).other_methods`). You may need to update your notebooks for this.

This is the current raven conda build: https://anaconda.org/Zeitsperre/raven/files (using Raven-rev312.zip). If this needs to be updated, let me know.

**Comment by Zeitsperre** [comment]
The reason why the docs are failing is that we are set to fail on warnings. This line is raising one:
```
checking consistency... /home/docs/checkouts/readthedocs.org/user_builds/pavics-raven/checkouts/367/docs/source/notebooks/Comparing_hindcasts_and_ESP_forecasts.ipynb: WARNING: document isn't included in any toctree
```
All we need to do is add that notebook to the docs, and it should build fine.

**Comment by huard** [comment]
@richardarsenault As long as you have ravenpy installed from master (which includes your forecasting changes) it should work. I didn't have to reinstall RavenC or Ostrich. 
What kind of errors messages do you get ? 
Anyway, Ok, I'll merge. Christian and I have pretty much finished updating RavenPy with GR4JCN semi-distributed. This comes with a number of changes that need to be propagated to RavenWPS.

**Comment by richardarsenault** [comment]
Basically pytest doesn't even run in ravenWPS, says package metalink is not found. But I can't seem to conda install metalink nor can I pip install metalink, so that's where I am at now!

I have 3 failing tests in ravenpy, 2 of which are decimal precision assert errors, one is an error in the GR4J routing: TestHydroRouting:test_hydro_routing_upstream_ids --> if level is None and by is None: error" You need to supply one of 'by' and 'level'.

Thanks, I'll try and reset from master now.

**Comment by huard** [comment]
"pymetalink"

**Comment by Zeitsperre** [comment]
That's really strange. `pymetalink` is in both the environment.yml and requirements.txt. It should be installed?

**Pull request merged**
